File: valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        openers = '({['
        closers = ')}]'
        stack = []
        for ch in s:
            if ch in openers:
                stack.append(ch)
            else:
                if len(stack) > 0 and closers.index(ch) == openers.index(stack[-1]):
                    stack.pop()
                else:
                    return False
        return len(stack) == 0

File: valid-parentheses/valid-parentheses.cpp
// If opening parentheses, push to stack. Else, check if current closing parentheses matches opening parentheses at stack top. If so, pop. If not, return false. At the end of the string, if there are still items in the stack, return false, because these are opening parentheses that wouldn't be closed.
// Time Complexity  - O(n)
// Space Complexity - O(n)
​
class Solution {
public:
    bool isValid(string str) {
        
        stack <char> s;
        string openers = "({[";
        string closers = ")}]";
        
        for(auto ch : str) {
            if(openers.find(ch) != string::npos)
                s.push(ch);
            else {
                if(s.empty()) return false;
                if(openers.find(s.top()) == closers.find(ch))
                    s.pop();
                else return false;
            }
        }
        
        return s.empty();
    }
};
